scripts/generate_fish_completion.py

```diff
@@ -0,0 +1,293 @@
+_FISH_BAZEL_HEADER = """#!/usr/bin/env fish
+# Copyright 2020 The Bazel Authors. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# Fish completion for Bazel commands and options.
+#
+# This script was generated from a specific Bazel build distribution. See
+# https://github.com/bazelbuild/bazel/blob/master/scripts/generate_fish_completion.py
+# for details and implementation.
+"""
+_FISH_BAZEL_COMMAND_LIST_VAR = 'BAZEL_COMMAND_LIST'
+_FISH_BAZEL_SEEN_SUBCOMMAND = '__bazel_seen_subcommand'
+_FISH_BAZEL_SEEN_SUBCOMMAND_DEF = """
+function {} -d "Checks whether the current command line contains a bazel subcommand."
+    {} ${}
+end
+""".format(_FISH_BAZEL_SEEN_SUBCOMMAND, _FISH_SEEN_SUBCOMMAND_FROM,
+           _FISH_BAZEL_COMMAND_LIST_VAR)
+
+
+class BazelCompletionWriter(object):
+    """Constructs a Fish completion script for Bazel."""
+    def __init__(self, bazel, output_user_root):
+        """Initializes writer state.
+
+        Args:
+            bazel: String containing a path the a bazel binary to run.
+            output_user_root: String path to user root directory used for
+                running bazel commands.
+        """
+        self._bazel = bazel
+        self._output_user_root = output_user_root
+        self._startup_options = self._get_options_from_bazel(
+            ('help', 'startup_options'))
+        self._bazel_help_completion_text = self._get_bazel_output(
+            ('help', 'completion'))
+        self._param_types_by_subcommand = self._get_param_types()
+        self._subcommands = self._get_subcommands()
+
+    def write_completion(self, output_file):
+        """Writes a Fish completion script for Bazel to an output file.
+
+        Args:
+            output_file: File object opened in a writable mode.
+        """
+        output_file.write(_FISH_BAZEL_HEADER)
+        output_file.write('set {} {}\n'.format(
+            _FISH_BAZEL_COMMAND_LIST_VAR,
+            ' '.join(c.name for c in self._subcommands)))
+        output_file.write(_FISH_BAZEL_SEEN_SUBCOMMAND_DEF)
+        for opt in self._startup_options:
+            opt.write_completion(output_file)
+        for sub in self._subcommands:
+            sub.write_completion(output_file)
+
+    def _get_bazel_output(self, args):
+        return subprocess.check_output(
+            (self._bazel, '--output_user_root={}'.format(
+                self._output_user_root)) + tuple(args),
+            text=True)
```

Comment:
Can you avoid using `text` here to keep compatibility with older python versions?